runtime/opt/taupage/init.d/04-configure-network-interfaces.py

```diff
@@ -0,0 +1,244 @@
+            wait_for_network_interface_attachment(device_index)
+        except Exception as e:
+            logging.exception(e)
+            sys.exit(3)
+        logging.info("Attached interface '%s'", name)
+
+
+def process_arguments():
+    parser = argparse.ArgumentParser(
+        description='Prepares interfaces according to the description in /meta/taupage.yaml'
+    )
+
+    parser.add_argument(
+        '-f',
+        '--file',
+        dest='filename',
+        default='/meta/taupage.yaml',
+        help='config file in YAML'
+    )
+
+    parser.add_argument(
+        '-d',
+        '--debug',
+        action='store_true',
+        help='log additional info, for debugging purposes'
+    )
+
+    parser.add_argument(
+        '-r',
+        '--region',
+        dest='region',
+        help='uses a specific AWS region instead of querying the instance metadata'
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = process_arguments()
+
+    # Setup logging
+    if args.debug:
+        configure_logging(logging.DEBUG)
+    else:
+        configure_logging(logging.INFO)
+
+    current_region = args.region if args.region else detect_region()
+
+    # Load configuration from YAML file
+    config = get_config(args.filename)
+
+    if config.get("network_interfaces"):
+        handle_network_interfaces(current_region, config)
+
+        # The goal here is to be able to assign static IPs to instances
+        # Within the Zalando AWS account setup we have a private subnet per
+        # AZ. The idea is to create an ENI in each AZ in the private subnet where
+        # you want a static IP. This means, your instance is going to have two
+        # network interfaces on the same subnet, which causes some issues.
+        #
+        # The below code is based of the explaination at: https://goo.gl/2D8KrV
+        # for handling two network interfaces in the same subnet.
+
+        # Setting this to 1 Allows you to have multiple network interfaces on the same
+        # subnet, and have the ARPs for each interface be answered based
+        # on whether or not the kernel would route a packet from the
+        # the ARP'd IP out that interface
+        with open("/proc/sys/net/ipv4/conf/all/arp_filter", "w") as all_arp_filter:
+            all_arp_filter.write("1")
+        network_interfaces = []
+        default_gateway = netifaces.gateways()['default'][netifaces.AF_INET][0]
+
+        for device_index in range(0, len(config.get("network_interfaces")) + 1):
+            network_interfaces.append("eth{}".format(device_index))
+
+        # Run dhclient on all newly created interfaces to enable them to get IPs
+        # Note, we do not run dhclient on eth0 as this may affect network connectivity
+        # of the instance
+        for network_interface in network_interfaces[1:]:
+            call_command(["dhclient", str(network_interface)], allowed_error_codes=[0, 2])
```

Comment:
Why is exit status 2 fine? Is it documented anywhere?

Reply:
You've spotted a bug here :) this was copied over from the route add command, where an exit code of 2 is OK as it means the route already exists.
PR updated.

Reply:
Could you link the relevant documentation? I couldn't find the exit statuses of `ip route`/`ip rule` documented anywhere at a first glance.

Besides, does it really make sense to just blindly proceed with the next commands instead of failing? The script will only run once during the boot process, so I'd expect it to either configure the instance without any issues or fail and log an error (with the instance being terminated and restarted automatically). The current approach seems more brittle to me — some commands will work, some will silently break but then we'll run the next commands anyway, leaving the instance in an indeterminate state. Are you sure this is absolutely required?

Reply:
> Could you link the relevant documentation? I couldn't find the exit statuses of ip route/ip rule documented anywhere at a first glance.

I copied this over from @mfelsche work, he may be able to comment on it, I too was unable to find docs relating too it. I did test it out manually though.

> Besides, does it really make sense to just blindly proceed with the next commands instead of failing? The script will only run once during the boot process, so I'd expect it to either configure the instance without any issues or fail and log an error (with the instance being terminated and restarted automatically)

Agreed with this approach, it makes sense to me to adjust to only accept exit code 0.